docs/manpages/tracker-unique.1
.TH tracker-unique 1 "Oct 2008" GNU "User Commands"

.SH NAME
tracker-unique \- command line tool to ask for unique values of a
property in tracker.

.SH SYNOPSIS
\fBtracker-unique\fR 
[OPTIONS]
-p \fIRDFFILE\fR \fIFIELD\fR [FIELD...]

.SH DESCRIPTION
.B tracker-unique
gets all the unique values of \fIFIELD\fR
after running the RDF query in \fIRDFFILE\fR. 

.PP
More than one field can be used, and then it will return the 
unique values of the combination.

.SH OPTIONS
.TP
.B \-?, \-\-help
Show summary of options.
.TP
.B \-p, \-\-path=RDFFILE
A file containing a RDF query. (See
documentation for description of the format).
.TP
.B FIELD
A valid property in tracker. For more information on tracker
properties, see
.BR tracker-services (1)
.TP
.B \-s, \-\-service=SERVICE
Limit the query to the specified service type. For more information on
tracker services, see 
.BR tracker-services (1)
.TP
.B \-c, \-\-count=FIELD
Count different values of \fIFIELD\fR in the result of the query.
.TP
.B \-u, \-\-sum=FIELD
Add all values of \fIFIELD\fR in the result of the query.
.TP
.B \-o, \-\-desc
Sort to descending order


.SH SEE ALSO
.BR trackerd (1),
.BR tracker-services (1).

.SH AUTHOR
tracker-unique was written by Mikael Ottela <mikael.ottela@ixonos.com>.


NEWS
Tracker 0.6.6
============= (03 Mar 2008)

  Applet
  * HIG fix for preference window.
  * Progress for index merges added.

  Tracker daemon
  * Fixed shutdown of daemon when paused to prevent hangs.
  * Fixed an unfinished final index merge step when next restarted.



Tracker 0.6.5
============= (27 Feb 2008)

  Applet
  * Removed popup window and replaced with tooltip displaying status
    and progress.
  * New Applet preferences to turn off animation/hide icon/set
    smart pause.
  * Smart pause feature which can auto-pause the Tracker daemon
    whenever the user presses a key or moves the mouse, thereby
    improving system responsiveness during indexing.
  * Annoying merge messages have now been removed.
  * Reindex option added.
  * About box added.
  * Now hidden if indexing is disabled.
  * Bugs with constant animation fixed.
  * Animation rate slowed right down so as to be less annoying.

  Tracker daemon
  * Temporary blacklisting of frequently changing files (they are
    indexed after one hour or on next restart of Tracker daemon).
    This prevents constant indexing from downloads and torrents.
  * More blacklisting of common tmp formats and vmware files.
  * Memory leak fixes.
  * Removed CPU-heavy corruption scan which can last long time and
    occurs on battery too at startup when tracker daemon was not
    previously shut down properly. Instead we now use bigger
    transactions with fsync on to prevent corruption in the first
    place (only slightly more disk IO is used during indexing as a
    result).
  * Fixed crashers.
  * Fixed D-Bus shutdown/reindex call.



Tracker 0.6.4
============= (11 Dec 2007)

  * Made indexing more robust by pausing if disk space is low or
    index grows too big.
  * Limit log file size to 10MB.
  * New Tracker Applet - animates when indexing, provides ability to
    pause indexing as well as viewing status and progress feedback
    from indexer, statistics, and notification warnings; it can
    also launch Tracker Preferences and Tracker Search Tool.
  * New power management options enable much better customization.
  * Ignored files fixes.
  * Deskbar/tracker integration fixes.
  * Made most prefs live and affect Tracker in real time; others
    will prompt for restart and/or reindex where necessary.
  * Shell script fixes
  * Fixed Imap bug with embedded Auth in URI.
  * Built-in corruption check and scan when Tracker is not shut down
    cleanly - prevents infinite looping.
  * Fix index deletions.
  * Many more bug fixes and stability improvements.



Tracker 0.6.3
============= (25 Sep 2007)

  * Fixed memory leaks.
  * New index merging which allows for much better performance and
    much less disk hogging I/O.
  * Battery throttling - Tracker now stops indexing when on battery.
  * Many more optimisations - transactions, use of O_NOATIME and
    posix_fadvise everywhere.
  * Check files exist before including in search results.
  * Fixed disabling of indexing option in tracker-preferences.
  * HIG fixes to tracker-preferences screen.
  * Fixed resetting of stats.

  note:
    This version will cause your hard drive to be re-indexed so
    that stats are fixed (previous version broke them).



Tracker 0.6.2
============= (25 Sep 2007)

  * New Sqlite-based indexer which utilises the new incremental blob
    I/O in sqlite 3.4.
  * Highly-optimised email indexing (up to 5x faster).
  * Dramatically reduced disk access and disk contention.
  * Indexer now pauses for a grace period when non-Tracker processes
    write to disk (providing changed files are being watched by
    Tracker) - minimises slowdowns when compiling or checking out
    source code.
  * Makes use of idle class disk IO scheduling if available.
  * Makes preliminary use of NO_ATIME (some disk access still uses
    fopen).
  * Fixed restore of user metadata on re-index (keywords are
    auto-restored).
  * Added increased number of (junk) files to automatically ignore.
  * Improved stopwords.
  * New Deskbar handler that uses the new Deskbar api (2.19+).
  * Fixed old Deskbar handler to remove race condition causing
    crashes.
  * Fixed a number of annoying bugs in email indexing and Tracker
    Preferences.

  note:
    This version will cause your hard drive to be re-indexed due to
    the new Sqlite indexer backend.


Tracker 0.6.1
============= (08 Aug 2007)

  * translations and bugfixes


Tracker 0.6.0 "better late than never"
============= (24 Jul 2007)

  * Evolution Emails (mbox/pop/imap) now indexed.
  * Gaim/Pidgin chat logs now indexed.
  * Applications now indexed.
  * New files indexed - all spreadsheets, csv (these require
    Gnumeric to be installed) and oo draw.
  * Vastly improved and new search UI using a sidebar for categories
    and a new metadata tile.
  * Tagging support in the UI.
  * New tracker-preferences UI for easy setting of prefs.
  * Added libtracker-gtk widget toolkit.
  * Now uses XDG directories.
  * Optimised triple store.
  * Handles multiple triple stores (files, emails and user data are
    all separate dbs).
  * New IDF based ranking algorithm.
  * XMP metadata support added.
  * Made use of async calls in tracker-search-tool to prevent it
    from freezing.
  * Added a 60 second delay before indexing to prevent slowdowns on
    desktop start-up.
  * Added support for crawling directories without watching overhead
    (useful for indexing source code without causing slowdowns when
     compiling).


Tracker 0.5.4 "Speed Daemon"
============= (26 Jan 2007)

  * Indexing at ludicrous speed - massively optimised indexing so
    its now 10x faster than previous version. Indexing speeds are
    now around 100 text files per second (which is about the
    maximum possible considering the I/O time to read 100 files from
    a hard drive).
  * Index while you work - now scheduled even better so it should
    never slow the system down whilst allowing users to work without
    being affected by or even noticing it.

src/tracker-utils/tracker-unique.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/*
 * Copyright (C) 2006, Mr Jamie McCracken (jamiemcc@gnome.org)
 * Copyright (C) 2008, Nokia
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include "config.h"

#include <sys/param.h>
#include <stdlib.h>
#include <locale.h>

#include <glib.h>
#include <glib/gi18n.h>

#include <libtracker/tracker.h>

#ifdef G_OS_WIN32
#include <trackerd/mingw-compat.h>
#endif /* G_OS_WIN32 */

static gchar	    **fields;
static gchar	     *service;
static gchar	     *path;
static gchar	     *count;
static gchar         *sum;
static gboolean       descending;

static GOptionEntry   entries[] = {
	{ "path", 'p', 0, G_OPTION_ARG_STRING, &path,
	  N_("Path to use in query"),
	  NULL
	},
	{ "service", 's', 0, G_OPTION_ARG_STRING, &service,
	  N_("Search from a specific service"),
	  NULL
	},
	{ "count", 'c', 0, G_OPTION_ARG_STRING, &count,
	  N_("Count instances of unique fields of this type"),
	  "e.g. File:Mime"
	},
	{ "sum", 'u', 0, G_OPTION_ARG_STRING, &sum,
	  N_("Sum the values of this field"),
	  "e.g. File:Mime"
	},
	{ "desc", 'o', 0, G_OPTION_ARG_NONE, &descending,
	  N_("Sort to descending order"),
	  NULL},
	{ G_OPTION_REMAINING, 0, 0,
	  G_OPTION_ARG_STRING_ARRAY, &fields,
	  N_("Required fields"), NULL},
	{ NULL }
};

static void
get_meta_table_data (gpointer value)
{
	gchar **meta;
	gchar **p;
	gint	i;

	meta = value;

	for (p = meta, i = 0; *p; p++, i++) {
		if (i == 0) {
			g_print ("  %s", *p);
		} else {
			g_print (", %s", *p);
		}
	}

	g_print ("\n");
}

int
main (int argc, char **argv)
{
	TrackerClient	*client;
	ServiceType	 type;
	GOptionContext	*context;
	GError		*error = NULL;
	gchar		*content;
	gchar		*buffer = NULL;
	gsize		 size;
	GPtrArray	*array;

	setlocale (LC_ALL, "");

	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	context = g_option_context_new ("- Get unique values with an optional RDF query filter");
	g_option_context_add_main_entries (context, entries, NULL);
	g_option_context_parse (context, &argc, &argv, NULL);

	if (!fields) {
		gchar *help;

		g_printerr ("%s\n\n",
			    _("Fields are missing"));

		help = g_option_context_get_help (context, TRUE, NULL);
		g_option_context_free (context);
		g_printerr (help);
		g_free (help);

		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	client = tracker_connect (FALSE);

	if (!client) {
		g_printerr ("%s\n",
			    _("Could not establish a DBus connection to Tracker"));
		return EXIT_FAILURE;
	}

	if (!service) {
		g_print ("%s\n",
			 _("Defaulting to 'files' service"));

		type = SERVICE_FILES;
	} else {
		type = tracker_service_name_to_type (service);

		if (type == SERVICE_OTHER_FILES && g_ascii_strcasecmp (service, "Other")) {
			g_printerr ("%s\n",
				    _("Service not recognized, searching in other files..."));
		}
	}

	if (path) {
		gchar *path_in_utf8;

		path_in_utf8 = g_filename_to_utf8 (path, -1, NULL, NULL, &error);
		if (error) {
			g_printerr ("%s:'%s', %s\n",
				    _("Could not get UTF-8 path from path"),
				    path,
				    error->message);
			g_error_free (error);
			tracker_disconnect (client);

			return EXIT_FAILURE;
		}

		g_file_get_contents (path_in_utf8, &content, &size, &error);
		if (error) {
			g_printerr ("%s:'%s', %s\n",
				    _("Could not read file"),
				    path_in_utf8,
				    error->message);
			g_error_free (error);
			g_free (path_in_utf8);
			tracker_disconnect (client);

			return EXIT_FAILURE;
		}

		g_free (path_in_utf8);

		buffer = g_locale_to_utf8 (content, size, NULL, NULL, &error);
		g_free (content);

		if (error) {
			g_printerr ("%s, %s\n",
				    _("Could not convert query file to UTF-8"),
				    error->message);
			g_error_free (error);
			tracker_disconnect (client);

			return EXIT_FAILURE;
		}
	}

	array = tracker_metadata_get_unique_values_with_count_and_sum (client,
								       type,
								       fields,
								       buffer,
								       count,
								       sum,
								       descending,
								       0,
								       512,
								       &error);
	g_free (buffer);

	if (error) {
		g_printerr ("%s, %s\n",
			    _("Could not query search"),
			    error->message);
		g_error_free (error);

		return EXIT_FAILURE;
	}

	if (!array) {
		g_print ("%s\n",
			 _("No results found matching your query"));
	} else {
		g_print ("%s:\n",
			 _("Results"));

		g_ptr_array_foreach (array, (GFunc) get_meta_table_data, NULL);
		g_ptr_array_free (array, TRUE);
	}

	tracker_disconnect (client);

	return EXIT_SUCCESS;
}
